util/llbutil/secretprovider/aws_creds.go

```diff
@@ -0,0 +1,116 @@
+package secretprovider
+
+import (
+	"context"
+	"net/url"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/moby/buildkit/session/secrets"
+	"github.com/pkg/errors"
+	"gopkg.in/ini.v1"
+)
+
+// AWSCredentialProvider can load AWS settings from the environment.
+type AWSCredentialProvider struct {
+	mu          sync.Mutex
+	config      *ini.File
+	credsConfig *ini.File
+}
+
+// NewAWSCredentialProvider creates and returns a credential provider for AWS.
+func NewAWSCredentialProvider() *AWSCredentialProvider {
+	return &AWSCredentialProvider{}
+}
+
+// GetSecret attempts to find an AWS secret from either the environment or a local config file.
+func (c *AWSCredentialProvider) GetSecret(ctx context.Context, name string) ([]byte, error) {
+
+	names := map[string]int{
+		"AWS_ACCESS_KEY_ID":     0,
+		"AWS_SECRET_ACCESS_KEY": 0,
+		"AWS_SESSION_TOKEN":     0,
+		"AWS_DEFAULT_REGION":    0,
+	}
+
+	q, err := url.ParseQuery(name)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to parse secret info")
+	}
+
+	secretName := q.Get("name")
+
+	if _, ok := names[secretName]; ok {
+		if val, ok := c.loadFromEnv(ctx, secretName); ok {
+			return []byte(val), nil
+		}
+
+		val, ok, err := c.loadFromConfig(ctx, secretName)
+		if err != nil {
+			return nil, err
+		}
+
+		if ok {
+			return []byte(val), nil
+		}
+	}
+
+	return nil, secrets.ErrNotFound
+}
+
+func (c *AWSCredentialProvider) loadFromEnv(ctx context.Context, name string) (string, bool) {
+	return os.LookupEnv(name)
+}
+
+func (c *AWSCredentialProvider) loadFromConfig(ctx context.Context, name string) (string, bool, error) {
```

Comment:
Have we discussed if anything else is needed to support assuming different roles, or is that outside of this use-case?

If we need to support that, I wonder if there's an aws-specific library that we should be using instead of reading directly from the ini.

Reply:
IIRC it's not easy to access the credentials using the Go lib once they've been parsed. I'll take another look. I think the goal for the first version was to support the envs & `~/.aws` in a limited fashion so perhaps we can iterate on other profiles, etc.

Reply:
Can we use [LoadDefaultConfig](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/config#LoadDefaultConfig) or do we need to load the values in a different way?

Reply:
I took another look at this. I was initially wary of using the config loader because it hits several different sources and it's not as configurable as it should be IMO.

Unfortunately, it doesn't seem possible to limit the sources to envs and standard config files only. It may also hit other AWS APIs to retrieve some settings.

However, I think it's alright to use this approach as it will likely improve the overall UX (as in it will discover settings from the environment without much user consideration).

The one feature I did disable manually is the EC2 metadata remote source for config. This seems rather peripheral in terms of features we need to support. Plus it generates some ugly errors.  

Example:
```
earthfile2llb group error: async force execution for +basic: unlazy force execution: failed to load AWS credentials: failed to refresh cached credentials, no EC2 IMDS role found, operation error ec2imds: GetMetadata, access disabled to EC2 IMDS via client option, or "AWS_EC2_METADATA_DISABLED" environment variable
```

CC @vladaionescu @alexcb

Reply:
> It may also hit other AWS APIs to retrieve some settings.

I think this is fine? I wouldn't be surprised if their MFA would require something like this?

> Plus it generates some ugly errors.
Maybe we could add to this switch statement under
https://github.com/earthly/earthly/blob/bafb458fda83a990c35d16038bc698497e7bf70e/cmd/earthly/app/run.go#L185 to display these errors in a more user-friendly way?

Reply:
I believe MFA is usually run upfront and will save credentials in the environment or config files.